axlearn/common/attention_bias.py

```diff
@@ -513,29 +519,15 @@ def _bool_value(self) -> Optional[Tensor]:
         Shape: [batch, target_len, source_len].
 
         Raises:
-            NotImplementedError. If `target_positions.ndim not in [1,2]`.
+            ValueError. If `(target|source)_positions.ndim not == 2`.
         """
-        target_positions, source_positions = jnp.indices(self.shape, sparse=True)
-        # Shape: [1, target_len, 1], [1, 1, source_len].
-        target_positions, source_positions = target_positions[None], source_positions[None]
-        if self.target_positions is not None:
-            target_positions = self.target_positions
-            if target_positions.ndim not in [1, 2]:
-                raise NotImplementedError(f"Shape of target_positions: {target_positions.shape}.")
-            if target_positions.ndim == 1:
-                # Shape: [batch, 1] + [target_len] = [batch, target_len]
-                # pylint: disable-next=unsubscriptable-object
-                target_positions = target_positions[:, None] + jnp.arange(self.shape[0])
-            elif target_positions.ndim == 2:
-                shape_with_batch_dim = (1, *self.shape)
-                # Raise an exception if shapes aren't compatible. We don't use the output.
-                jnp.broadcast_shapes(
-                    (target_positions.shape[0], 1, target_positions.shape[1]), shape_with_batch_dim
-                )
-            else:
-                raise NotImplementedError(f"Invalid value {target_positions.ndim=}.")
-            target_positions = target_positions[..., None]  # Shape: [batch, target_len, 1].
-
+        target_positions, source_positions = self.target_positions, self.source_positions
+        if target_positions.ndim != source_positions.ndim != 2:
+            raise ValueError(
+                f"{target_positions.shape=} or {source_positions.shape=} is not rank 2."
+            )
+        target_positions = einops.rearrange(target_positions, "b t -> b t 1")
```

Comment:
fly-by comment: I wonder if we should do the same without introducing extra `einops` dependency.

Reply:
We could use the following code:
```
target_positions = target_positions[:, :, None]
```

However, I prefer using einops because it makes the meaning of shape transformations clearer in the code.

Since Axlearn and Ajax already have einops installed by default, we are increasingly using it.
Additionally, the G-model codebase also heavily relied on einops.

Would you prefer not to use einops?

Reply:
I don't have a strong opinions - just from engineering perspective we might want to prefer native approach over 3rd party dependencies in case of unexpected surprises. But you can make the call if you think `einops` is stable enough.

Reply:
Thank you for your concern.
I generally agree with your opinion. einops improves code readability, which I believe will contribute to maintaining a healthy codebase in the long run.
Also, as far as I remember, Google uses it in production, so I consider it to be a reliable thirdparty.

Reply:
> Google uses it in production

fwiw I didn't see it in production codebase, but maybe "uses it in production" is a vague term after all.